internal/tasks/manifests.go

```diff
@@ -891,3 +894,160 @@ func (j *Janitor) setManifestAndParentsToPending(ctx context.Context, manifestDi
+}
+
+func (j *Janitor) processTrivySecurityInfo(tx *gorp.Transaction, securityInfo keppel.TrivySecurityInfo, labels prometheus.Labels) error {
+	//load corresponding repo, account and manifest
+	repo, err := keppel.FindRepositoryByID(tx, securityInfo.RepositoryID)
+	if err != nil {
+		return fmt.Errorf("cannot find repo for manifest %s: %s", securityInfo.Digest, err.Error())
+	}
+	account, err := keppel.FindAccount(tx, repo.AccountName)
+	if err != nil {
+		return fmt.Errorf("cannot find account for repo %s: %s", repo.FullName(), err.Error())
+	}
+	manifest, err := keppel.FindManifest(tx, *repo, securityInfo.Digest)
+	if err != nil {
+		return fmt.Errorf("cannot find manifest for repo %s and digest %s: %s", repo.FullName(), securityInfo.Digest, err.Error())
+	}
+
+	err = j.doSecurityCheck(*account, *repo, *manifest, &securityInfo)
+	if err != nil {
+		return err
+	}
+	_, err = tx.Update(&securityInfo)
+	if err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func (j *Janitor) doSecurityCheck(account keppel.Account, repo keppel.Repository, manifest keppel.Manifest, securityInfo *keppel.TrivySecurityInfo) (returnedError error) {
+	//clear timing information (this will be filled down below once we actually talk to Trivy;
+	//if any preflight check fails, the fields stay at nil)
+	securityInfo.CheckedAt = nil
+	securityInfo.CheckDurationSecs = nil
+
+	//skip validation while account is in maintenance (maintenance mode blocks
+	//all kinds of activity on an account's contents)
+	if account.InMaintenance {
+		securityInfo.NextCheckAt = j.timeNow().Add(j.addJitter(1 * time.Hour))
+		return nil
+	}
+
+	continueCheck, err := j.checkPreConditionsForTrivy(account, repo, manifest, securityInfo)
+	if err != nil {
+		return err
+	}
+	if !continueCheck {
+		return nil
+	}
+
+	//we know that this image will not be "Unsupported", so the rest is the part where we actually
+	//talk to Trivy (well, mostly anyway), so that part deserves to be measured for performance
+	checkStartedAt := j.timeNow()
+	defer func() {
+		if returnedError == nil {
+			checkFinishedAt := j.timeNow()
+			securityInfo.CheckedAt = &checkFinishedAt
+			duration := checkFinishedAt.Sub(checkStartedAt).Seconds()
+			securityInfo.CheckDurationSecs = &duration
+		}
+	}()
+	//also we don't allow Trivy to take more than 10 minutes on a single image (which is already an
+	//insanely generous timeout)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer cancel()
+
+	//ask Trivy for the security status of the manifest
+	securityInfo.Message = "" //unless it gets set to something else below
+	imageRef := keppel.ImageReference{
+		Host:      j.cfg.APIPublicHostname,
+		RepoName:  repo.Name,
+		Reference: keppel.ManifestReference{Digest: manifest.Digest},
+	}
+	requestURL := j.cfg.Trivy.URL
+	requestURL.Path = "/trivy"
+	requestURL.RawQuery = "image=" + imageRef.String()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL.String(), http.NoBody)
```

Comment:
That's a response, not a request.
```suggestion
	resp, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL.String(), http.NoBody)
```

Reply:
The request is only done in the code I add in the next force push.